Docs/309 actions sessions and cookies (#132)
hanami/guides: content/v2.0/actions/cookies.md

---
title: Cookies
order: 80
---

Actions can set [cookies](https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies) on outgoing requests via the `response` object.

```ruby
module Bookshelf
  module Actions
    module Books
      class Index < Bookshelf::Action
        def handle(request, response)
          response.cookies["tasty_cookie"] = "strawberry"
        end
      end
    end
  end
end
```

> Cookies subsequently sent by the browser can be read from the request.

Reviewer: ```suggestion
Cookies subsequently sent by the browser can be read from the `request`.
```

Reviewer: Just to match `response` in the line above.

```ruby
module Bookshelf
  module Actions
    module Books
      class Index < Bookshelf::Action
        def handle(request, response)
          request.cookies["tasty_cookie"] # => "strawberry"
        end
      end
    end
  end
end
```

## Cookie configuration

You can set one or more of the following options for cookies issued by actions using the action cookies config on your app.

  * `:domain` - `String` (`nil` by default), the domain
  * `:path` - `String` (`nil` by default), a relative URL
  * `:max_age` - `Integer` (`nil` by default), cookie duration expressed in seconds
  * `:secure` - `Boolean` (`true` by default if using SSL), restrict cookies to secure connections
  * `:httponly` - `Boolean` (`true` by default), restrict JavaScript access to cookies

```ruby
# config/app.rb

module Bookshelf
  class App < Hanami::App
    config.actions.cookies = {
      domain: "hanami.example.com",
      secure: true,
      httponly: true,
      path: "/foo",
      max_age: 300
    }
  end
end
```

> This configuration can be overridden on an as needs basis passing a hash, which has a `value` key representing the value of the cookie, and any properties to override.

Reviewer: ```suggestion
This configuration can be overridden within an action passing a hash, which has a `value` key representing the value of the cookie, and any properties to override.
```

"on an as needs basis" just took me a couple reads to fully grok
